File: src/main/java/com/github/cao/awa/conium/item/event/use/ConiumItemUseOnBlockEvent.java
package com.github.cao.awa.conium.item.event.use;

import com.github.cao.awa.conium.event.type.ConiumEventArgType;
import com.github.cao.awa.conium.event.context.ConiumEventContext;
import com.github.cao.awa.conium.event.context.ConiumEventContextBuilder;
import com.github.cao.awa.conium.item.event.ConiumItemEvent;
import com.github.cao.awa.conium.parameter.ParameterSelective;
import com.github.cao.awa.conium.parameter.ParameterSelective1;
import com.github.cao.awa.conium.parameter.ParameterSelective2;
import net.minecraft.item.Item;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.server.world.ServerWorld;
import org.jetbrains.annotations.NotNull;

public class ConiumItemUseOnBlockEvent extends ConiumItemEvent<ParameterSelective2<ServerWorld, ItemUsageContext>> {
    @NotNull
    @Override
    public ConiumEventContext<? extends ParameterSelective> requirement() {
        return ConiumEventContextBuilder.requires(
                ConiumEventArgType.SERVER_WORLD,
                ConiumEventArgType.ITEM_USAGE_CONTEXT
        ).trigger(
                (identity, world, context) -> hasAny(identity, handler -> handler.trigger(world, context))
        );
    }
}

File: src/main/java/com/github/cao/awa/conium/item/event/use/ConiumItemUseOnBlockEvent.kt
package com.github.cao.awa.conium.item.event.use

import com.github.cao.awa.conium.event.context.ConiumEventContext
import com.github.cao.awa.conium.event.context.ConiumEventContextBuilder.requires
import com.github.cao.awa.conium.event.type.ConiumEventArgTypes
import com.github.cao.awa.conium.event.type.ConiumEventType
import com.github.cao.awa.conium.item.event.ConiumItemEvent
import com.github.cao.awa.conium.parameter.ParameterSelective
import com.github.cao.awa.conium.parameter.ParameterSelective2
import net.minecraft.item.ItemUsageContext
import net.minecraft.server.world.ServerWorld

class ConiumItemUseOnBlockEvent : ConiumItemEvent<ParameterSelective2<Boolean, ServerWorld, ItemUsageContext>>() {
    override fun requirement(): ConiumEventContext<out ParameterSelective> {
        return requires(
            ConiumEventArgTypes.SERVER_WORLD,
            ConiumEventArgTypes.ITEM_USAGE_CONTEXT
        ).attach(
            forever(ConiumEventType.ITEM_USE_ON_BLOCK)
        ).arise { identity, world, context ->
            noFailure(identity) {
                it.arise(
                    world,
                    context
                )
            }
        }
    }
}
